docs/source/_notebooks/keras-image-classifiers.rst

```diff
@@ -33,56 +35,210 @@ To start out, lets load our image classifier and data.
     
-    # and load our sample image
-    image = 'imagenet-samples/cat_dog.jpg'  # TODO: replace this image with something more interesting
+    # check the input format
+    print(model.input_shape)
+    dims = model.input_shape[1:3] # -> (height, width)
+    print(dims)
+
+
+.. parsed-literal::
+
+    (None, 224, 224, 3)
+    (224, 224)
+
+
+We see that we need a numpy tensor of shape (batches, height, width,
+channels), with the specified height and width.
+
+Loading our sample image:
+
+.. code:: ipython3
+
+    # we start from a path / URI. 
+    # If you already have an image loaded, follow the subsequent steps
+    image = 'imagenet-samples/cat_dog.jpg'
     
-    dims = model.input_shape[1:3] # 224x224
-    doc = eli5.keras.image_from_path(image, image_shape=dims)
-    doc = mobilenet_v2.preprocess_input(doc) # mobilenetv2-specific preprocessing (this operation is in-place)
+    # this is the original "cat dog" image used in the Grad-CAM paper
+    # check the image with Pillow
+    im = Image.open(image)
+    print(type(im))
+    display(im)
+
+
+.. parsed-literal::
+
+    <class 'PIL.JpegImagePlugin.JpegImageFile'>
+
+
+
+.. image:: ../_notebooks/keras-image-classifiers_files/keras-image-classifiers_5_1.png
+
+
+We see that this image will need some preprocessing to have the correct
+dimensions! Let's resize it:
+
+.. code:: ipython3
+
+    # we could resize the image manually
+    # but instead let's use a utility function from `keras.preprocessing`
+    # we pass the required dimensions as a (height, width) tuple
+    im = keras.preprocessing.image.load_img(image, target_size=dims) # -> PIL image
+    print(type(im))
+    display(im)
+
+
+.. parsed-literal::
+
+    <class 'PIL.Image.Image'>
+
+
+
+.. image:: ../_notebooks/keras-image-classifiers_files/keras-image-classifiers_7_1.png
+
+
+Looking good. Now we need to convert the image to a numpy array.
+
+.. code:: ipython3
+
+    # we use a routine from `keras.preprocessing` for that as well
+    # we get a 'doc', an object almost ready to be inputted into the model
     
-    # looking good?
-    display(keras.preprocessing.image.array_to_img(doc[0]))
+    doc = keras.preprocessing.image.img_to_array(im) # -> numpy array
+    print(type(doc), doc.shape)
 
 
 .. parsed-literal::
 
-    Using TensorFlow backend.
+    <class 'numpy.ndarray'> (224, 224, 3)
+
+
+.. code:: ipython3
+
+    # dimensions are looking good
+    # except that we are missing one thing - the batch size
+    
+    # we can use a numpy routine to create an axis in the first position
+    doc = np.expand_dims(doc, axis=0)
+    print(type(doc), doc.shape)
+
 
+.. parsed-literal::
+
+    <class 'numpy.ndarray'> (1, 224, 224, 3)
+
+
+.. code:: ipython3
+
+    # one last thing
+    # `keras.applications` models come with their own input preprocessing function
+    # for best results, apply that as well
+    
+    # mobilenetv2-specific preprocessing
+    # (this operation is in-place)
```

Comment:
If this is in-place, I think it's better to remove assignment and have only ``mobilenet_v2.preprocess_input(doc)``

Reply:
https://github.com/TeamHG-Memex/eli5/pull/315/commits/5d805945fc7a2c8287a86c63e422a0f0b7efba18